Doc/library/asyncio-stack.rst

```diff
@@ -0,0 +1,139 @@
+.. currentmodule:: asyncio
+
+
+.. _asyncio-stack:
+
+===================
+Stack Introspection
+===================
+
+**Source code:** :source:`Lib/asyncio/stack.py`
+
+-------------------------------------
+
+asyncio has powerful runtime call stack introspection utilities
+to trace the entire call graph of a running coroutine or task, or
+a suspended *future*.
+
+.. versionadded:: 3.14
+
+
+.. function:: print_call_graph(*, future=None, file=None, depth=1)
+
+   Print the async call graph for the current task or the provided
+   :class:`Task` or :class:`Future`.
+
+   The function receives an optional keyword-only *future* argument.
+   If not passed, the current running task will be used. If there's no
+   current task, the function returns ``None``.
+
+   If the function is called on *the current task*, the optional
+   keyword-only ``depth`` argument can be used to skip the specified
+   number of frames from top of the stack.
+
+   If *file* is not specified the function will print to :data:`sys.stdout`.
+
+   **Example:**
+
+   The following Python code:
+
+   .. code-block:: python
+
+      import asyncio
+
+      async def test():
+         asyncio.print_call_graph()
+
+      async def main():
+         async with asyncio.TaskGroup() as g:
+            g.create_task(test())
+
+      asyncio.run(main())
+
+   will print::
+
+      * Task(name='Task-2', id=0x1039f0fe0)
+      + Call stack:
+      |   File 't2.py', line 4, in async test()
+      + Awaited by:
+         * Task(name='Task-1', id=0x103a5e060)
+            + Call stack:
+            |   File 'taskgroups.py', line 107, in async TaskGroup.__aexit__()
+            |   File 't2.py', line 7, in async main()
+
+   For rendering the call stack to a string the following pattern
+   should be used:
+
+   .. code-block:: python
+
+      import io
+
+      ...
+
+      buf = io.StringIO()
+      asyncio.print_call_graph(file=buf)
+      output = buf.getvalue()
+
+
+.. function:: capture_call_graph(*, future=None)
+
+   Capture the async call graph for the current task or the provided
+   :class:`Task` or :class:`Future`.
+
+   The function receives an optional keyword-only *future* argument.
+   If not passed, the current running task will be used. If there's no
+   current task, the function returns ``None``.
+
+   If the function is called on *the current task*, the optional
+   keyword-only ``depth`` argument can be used to skip the specified
+   number of frames from top of the stack.
+
+   Returns a ``FutureCallGraph`` named tuple:
```

Comment:
Why namedtuple and not a dataclass?

Reply:
* We don't need any functionality of dataclasses here
* Because historically we try to keep the number of dependencies in asyncio to the minimum to make sure we're not increasing its load time. If using dataclasses was necessary here I'd 100% use them, but... it's just not, a named tuple does the trick.

Reply:
> * Because historically we try to keep the number of dependencies in asyncio to the minimum to make sure we're not increasing its load time

See https://github.com/python/cpython/pull/124640#discussion_r1803579075

Reply:
Alright, I'm pushing an update to use dataclasses, Kumar is right, our ability to evolve this API freely outweighs the tiny import overhead. Moreover, I'm sure most asyncio users use dataclasses in their program anyway, it's an incredibly popular API.